mesa/experimental/devs/simulator.py

```diff
@@ -0,0 +1,265 @@
+import numbers
+from typing import Any, Callable
+
+from .eventlist import EventList, Priority, SimulationEvent
+
+
+class Simulator:
+    """The Simulator controls the time advancement of the model.
+
+    The simulator uses next event time progression to advance the simulation time, and execute the next event
+
+    Attributes:
+        event_list (EventList): The list of events to execute
+        time (float | int): The current simulation time
+        time_unit (type) : The unit of the simulation time
+        model (Model): The model to simulate
+
+
+    """
+
+    # TODO: add replication support
+    # TODO: add experimentation support
+
+    def __init__(self, time_unit: type, start_time: int | float):
+        # should model run in a separate thread,
+        # and we can then interact with start, stop, run_until, and step?
+        self.event_list = EventList()
+        self.time = start_time
+        self.time_unit = time_unit
+        self.model = None
+
+    def check_time_unit(self, time: int | float) -> bool:
+        ...
+
+    def setup(self, model: "Model") -> None:
+        """Setup the model to simulate
+
+        Args:
+            model (Model): The model to simulate
+
+        Notes:
+            The basic assumption of the simulator is that a Model has a model.setup method that sets up the
+            model.
+
+        """
+
+        self.event_list.clear()
+        self.model = model
+        model.setup()
+
+    def reset(self):
+        """Reset the simulator by clearing the event list and removing the model to simulate"""
+        self.event_list.clear()
+        self.model = None
+
+    def run(self, time_delta: int | float):
```

Comment:
I think the naming isn't optimal here. Run sounds very generic, but the condition is very specific to run for a specified amount of time. 

The implementation also isn't correct, because it calls step() until the current time is smaller than the end time. But the step just calls the next event, which might be later than the end time. This already happens in the benchmarks example, which run for 1 step too much.

Reply:
Agreed. Maybe `simulate_time` or `progress_time`?

Reply:
Happy to change the name, but not sure yet how. Basically this method will run the model for the provided time_delta. So, `run_for` or `progress_time_by`, or `simulate_for` all seem ok.

@Corvince, I am not sure I fully follow. It currently runs while the clock time (i.e., `self.time`) is smaller than `end_time`. So, yes, it will change the time until self.time equals end_time in case of ABMSimulator. This means that time_delta is inclusive. If you say `run(10)`, it will run until time equals 10 which is in fact 11 steps. I don't see an easy solution around this. 

A few options:
1. shift to using `eventlist.peak_ahead()` to check the timestamp of each next event before we execute this event. This creates a bit of overhead. 
2. Schedule the first step for next tick (rather than the current `schedule_event_now`). This implicitly means we shift from zero based to 1 based indexing because the first step happens at `time=1`.
3. solve it in simulator.step by not executing the event if the time becomes to large and instead reschedule this event for its specified time. It needs to be rescheduled because we have gotten it from the eventlist using pop.

Reply:
```Python
run_for(time_delta=something)
````
seems good

Reply:
In the latest update, I added run_for and run_until. This also fixes the problem of running one step too much.

Still need to check a few things, but its almost there.